pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -751,6 +882,127 @@ pub mod pallet {
 			};
 		}
 
+		pub(crate) fn calculate_collect(
+			loc: TrancheLocator<T::PoolId, T::TrancheId>,
+			order: UserOrder<T::Balance, T::EpochId>,
+			pool: PoolDetails<T::AccountId, T::CurrencyId, T::EpochId, T::Balance, T::InterestRate>,
+			end_epoch: T::EpochId,
+		) -> Result<OutstandingCollections<T::Balance>, DispatchError> {
+			// No collect possible in this epoch
+			if order.epoch == pool.current_epoch {
+				return Ok(OutstandingCollections {
+					payout_currency_amount: Zero::zero(),
+					payout_token_amount: Zero::zero(),
+					remaining_supply_currency: order.supply,
+					remaining_redeem_token: order.redeem,
+				});
+			}
+
+			// It is only possible to collect epochs which are already over
+			let parse_until_epoch = end_epoch.min(pool.last_epoch_executed);
+
+			// We initialize the outstanding collections to the ordered amounts, and then fill the payouts based on executions in the accumulated epochs
+			let mut outstanding = OutstandingCollections {
+				payout_currency_amount: Zero::zero(),
+				payout_token_amount: Zero::zero(),
+				remaining_supply_currency: order.supply,
+				remaining_redeem_token: order.redeem,
+			};
+
+			// Parse remaining_supply_currency into payout_token_amount
+			outstanding =
+				Self::parse_invest_executions(&loc, outstanding, order.epoch, parse_until_epoch)?;
```

Comment:
This is great, thanks for breaking things into self-suficient functions!

Now we are passing a mutable value, mutate it, and re-assign it. Comparing to the alternative of just passing a mutable reference, I might like this more since it's more "functional"-like and it's easier to test. But I wonder if the reference approach is more efficient, considering that the mutations occur within a loop.

I would say, leave it as is and once we implement benchmarking for this we go over that. It would be a good idea to leave a `TODO` to remind us of that.

Reply:
Great idea, just added the `TODO`. Thanks!

Reply:
Pleasure!